File: apps/e7_utils/plots.py
import plotly.graph_objects as go
import plotly.offline as offline
from apps.e7_utils.battle_manager import BattleManager
from apps.e7_utils.user_manager import User
import numpy as np

def make_rank_plot(battles: BattleManager, user: User, filtered_battles=None):
    battles_df = battles.to_dataframe()
    battles_df = battles_df.sort_values(by="time").reset_index(drop=True)

    marker_default_color = '#0df8fd'
    marker_filtered_color = '#ff9900'

    marker_mask = [marker_default_color for _ in range(len(battles_df))]
    if filtered_battles is not None:
        marker_mask = [marker_default_color if seq not in filtered_battles else marker_filtered_color for seq in battles_df['seq_num'].values ]
    
    x = list(battles_df.index.values)
    y = list(battles_df['scores_1'])

    fig = go.Figure()



    # Create the plot
    fig.add_trace(go.Scatter(
        x=x,
        y=y,
        mode='lines+markers',
        line=dict(color='#4f9293', width=2),
        customdata=np.stack((battles_df['time'].str[:10], battles_df['grades_1']), axis=-1),
        marker=dict(
            symbol='circle',
            size=4,
            color=marker_mask,
            # line=dict(width=0.75, color='#0df8fd')
        ),
        hovertemplate='Points: %{y}<br>' \
                      'Date: %{customdata[0]}<br>' \
                      'League: %{customdata[1]}<extra></extra>' # Custom hover text
    ))

    # Update layout for dark mode
    fig.update_layout(
        autosize=True,
        font_family="Roboto, Open Sans",
        title=dict(
            text=f"{user.name}'s RTA Point Plot",
            font=dict(size=24, color='#dddddd'), # Adjust title color
            xanchor = 'center',
            yanchor = 'top',
            y = 0.95,  # Adjust vertical position if needed
            x = 0.5,
        ),
        xaxis=dict(
            title=dict(text='Battle Number (Chronological)', font=dict(size=18, color='#dddddd')), # Adjust x-axis title color
            showgrid=True,
            gridcolor='#8d8d8d', # Darker grid lines
            zeroline=False,
            tickfont=dict(size=12, color='#dddddd'), # Adjust tick color
        ),
        yaxis=dict(
            title=dict(text='Victory Points', font=dict(size=18, color='#dddddd')), # Adjust y-axis title color
            showgrid=True,
            gridcolor='#8d8d8d', # Darker grid lines
            zeroline=True,
            zerolinecolor='#dddddd',
            zerolinewidth=2,
            tickfont=dict(size=12, color='#dddddd'), # Adjust tick color
        ),
        # plot_bgcolor='#c704b0', # Dark background for the plot area
        # paper_bgcolor='#62a832', # Dark background for the entire figure

        plot_bgcolor='#1e222d', # Dark background for the plot area
        paper_bgcolor='#1e222d', # Dark background for the entire figure
    )


    # Convert the plot to HTML
    plot_html = offline.plot(fig, include_plotlyjs='cdn', output_type='div', config={'responsive': True})
    return plot_html

File: static/assets/js/e7/plots.js
export function generateRankPlot(battles, user, filteredBattles = null, zoomFiltered = false) {
    // Sort battles chronologically by time
    // console.log("Creating plot HTML for:", JSON.stringify(battles));
    // console.log("received Filtered Battles:", JSON.stringify(filteredBattles));
    battles.sort((a, b) => a["Date/Time"].slice(0,10).localeCompare(b["Date/Time"].slice(0,10)));

    // if the user is not passed, default the username to the ID of the player
    if (!user) {
        user = {name: `UID: ${battles[0]["P1 ID"]}`}
    }

    const markerDefaultColor = '#0df8fd';
    const markerFilteredColor = '#ff9900';

    const x = battles.map((_, i) => i);
    const y = battles.map(b => b["P1 Points"]);

    const markerMask = [];
    const zoom = {
        startX: null,
        endX: null,
        startY: null,
        endY: null
    }

    const zoomYPadding = 50;
    const zoomXPadding = 0.5;

    // iterate through battles and build list to color filtered battles distinctly 
    // and determine the area to zoom on if needed
    for (let [idx, battle] of battles.entries()) {
        if (filteredBattles && battle["Seq Num"] in filteredBattles) {
            if (zoomFiltered === true) {
                zoom.startX = idx < zoom.startX || zoom.startX === null ? idx - zoomXPadding : zoom.startX;
                zoom.startY = battle["P1 Points"] < zoom.startY + zoomYPadding || zoom.startY === null ? battle["P1 Points"] - zoomYPadding : zoom.startY;
                zoom.endX = idx > zoom.endX || zoom.endX === null ? idx + zoomXPadding : zoom.endX;
                zoom.endY = battle["P1 Points"] > zoom.endY - zoomYPadding || zoom.endY === null ? battle["P1 Points"] + zoomYPadding : zoom.endY;
            }
            markerMask.push(markerFilteredColor);
        } else {
            markerMask.push(markerDefaultColor);
        }
    };

    const customdata = battles.map(b => [
        b["Date/Time"].slice(0,10), // date
        b["P1 League"]              // league
    ]);

    const trace = {
        x: x,
        y: y,
        mode: 'lines+markers',
        line: {
            color: '#4f9293',
            width: 2
        },
        marker: {
            symbol: 'circle',
            size: 4,
            color: markerMask
        },
        customdata: customdata,
        hovertemplate:
            'Points: %{y}<br>' +
            'Date: %{customdata[0]}<br>' +
            'League: %{customdata[1]}<extra></extra>'
    };

    const layout = {
        autosize: true,
        font: {
            family: 'Roboto, Open Sans'
        },
        title: {
            text: `${user.name}'s RTA Point Plot`,
            font: { size: 24, color: '#dddddd' },
            xanchor: 'center',
            yanchor: 'top',
            y: 0.95,
            x: 0.5
        },
        xaxis: {
            title: {
                text: 'Battle Number (Chronological)',
                font: { size: 18, color: '#dddddd' }
            },
            showgrid: true,
            gridcolor: '#8d8d8d',
            zeroline: false,
            tickfont: { size: 12, color: '#dddddd' },
            range: zoom.startX ? [zoom.startX, zoom.endX] : null
        },
        yaxis: {
            title: {
                text: 'Victory Points',
                font: { size: 18, color: '#dddddd' }
            },
            showgrid: true,
            gridcolor: '#8d8d8d',
            zeroline: true,
            zerolinecolor: '#dddddd',
            zerolinewidth: 2,
            tickfont: { size: 12, color: '#dddddd' },
            range: zoom.startY ? [zoom.startY, zoom.endY] : null
        },
        plot_bgcolor: '#1e222d',
        paper_bgcolor: '#1e222d'
    };

    const config = {
        responsive: true
    };

    // Generate HTML string
    const divId = `plot-${Math.random().toString(36).substr(2, 9)}`;
    const containerDiv = `<div id="${divId}"></div>`;
    const plotScript = `
<script>
    Plotly.newPlot('${divId}', [${JSON.stringify(trace)}], ${JSON.stringify(layout)}, ${JSON.stringify(config)});
</script>
`;

    return containerDiv + plotScript;
}